### DISCUSSION TITLE
the source term F = +U min(Q,0) in the momentum conserving method

### DISCUSSION THREAD

**Discussion by SIMAO75** [IBAMR/IBAMR#1483] [Q&A]
I am coding your momentum conserving, I have two questions:

1) what does the quantity Q in the term F = +U min(Q,0) correspond to?

2) What happens if we don't add this term in the source term.

Many thanks to you

**Comment by amneetb** [comment]
@SIMAO75 Can you point to the code that you are referring to?

**Comment by boyceg** [comment]
I think that this is a question about how the momentum equation is modified to account for internal fluid sources and sinks. If so, it is discussed in Appendix B of my PhD thesis (https://www.proquest.com/pagepdf/305469263).

**Comment by SIMAO75** [comment]
@amneetb I'm talking about this subroutine function navier_stokes_cons_source2d 
https://github.com/IBAMR/IBAMR/blob/master/src/navier_stokes/fortran/navier_stokes_divsource2d.f.m4

I turned the case of the dense sphere in translation (mass ratio 10^6), it works well in 2D with the CUI scheme but it diverges in 3D, I think that the stabilization term F = +U min(q,0) is interesting to put because I did not put it

**Comment by amneetb** [comment]
@SIMAO75 : To be on the same page are you talking about mass-momentum transport consistency for high-density ratio multiphase flows?

**Comment by SIMAO75** [comment]
@amneetb exactly i talek about t mass-momentum transport consistency for high-density ratio multiphase flows !

**Comment by amneetb** [comment]
Ok, we do not make use of source and sinks. `F = +U min(q,0)` is something from legacy in the INS class that is not used in INSVC class for multiphase flows. 

Recently, we have made some changes to the mass-momentum transport consistency classes. They have not propagated to the main branch. The main idea is that the auxiliary mass equation can be integrated by any integrator of choice and the momentum equation will take care of the mismatch between the integrators.  

Can you try the 3D code by selecting "SSPRK2" instead of "SSPRK3" for DENSITY_TS in the input file? For example:
https://github.com/IBAMR/IBAMR/blob/master/examples/multiphase_flow/ex6/input2d#L270

**Comment by amneetb** [comment]
> but it diverges in 3D,

BTW, are you talking about code in IBAMR or your own code for the 3D sphere case?

**Comment by SIMAO75** [comment]
@amneetb Thank you, but im talkin about my own code for the 3D sphere case

**Comment by amneetb** [comment]
I see. For us the 3D sphere case works as shown in the paper. But I think I may know what could be going on in your case.  

What integrators are you using for integrating density and for treating convective term in the momentum equation?

**Comment by SIMAO75** [comment]
@amneetb For the density transport I use a Range kutta of order 3 SPP-RK3 and the CUI scheme for the spatial derivative, the same scheme is used for the navier-stokes convective term, but it diverges in 3D

On the other hand when I use the WENO or CUbista scheme for the convective navier-stokes term the case works well in 3D

**Comment by amneetb** [comment]
Yep. Higher-order (HO) schemes are bit finicky when it comes to maintaining consistency. Can you make both HO integrators exactly the same?  That should fix it.  

Having the same integrator is not the ideal solution, because density equation is hyperbolic and momentum equation is parabolic. One would want to use different integrators for these two equations.  We have a way of addressing this problem, which we are testing on some problems. More on it later. For now try the above suggestion.

**Comment by SIMAO75** [comment]
Thanks for your suggestion, i will try it

**Comment by amneetb** [comment]
Hold on. Are you saying:

1) SSPRK3+CUI for density and SSPRK3+CUI for INS does not work. 
2) SSPRK3+CUI for density and WENO for INS works.
3) SSPRK3+CUI for density and SSPRK3+WENO (or SSPRK3+CUBISTA) for INS works.

**Comment by SIMAO75** [comment]
When i transport density and i deduse convectif term:
    SSPRK3+CUI for density and Euler 1+CUI INS does not work
    SSPRK3+weno for density and Euler 1+weno INS work.
    SSPRK3+Cubista for density and Euler 1+cubista INS work.

When i transport density and inetermediate momentum: 
    SSPRK3+CUI for density and SSPRK3 1+CUI INS does not work
    SSPRK3+weno for density and SSPRK3+weno INS work.
    SSPRK3+Cubista for density and SSPRK3+cubista INS work.

**Comment by amneetb** [comment]
> When i transport density and i deduse convectif term:
> 
>     SSPRK3+CUI for density and Euler 1+CUI INS does not work
> 
>     SSPRK3+weno for density and Euler 1+weno INS work.
> 
>     SSPRK3+Cubista for density and Euler 1+cubista INS work.
> 
> 
> 
> When i transport density and inetermediate momentum: 
> 
>     SSPRK3+CUI for density and SSPRK3 1+CUI INS does not work
> 
>     SSPRK3+weno for density and SSPRK3+weno INS work.
> 
>     SSPRK3+Cubista for density and SSPRK3+cubista INS work.

The first part makes sense because the two integrators are different for density and momentum. 

The second part also makes sense (the integrators are the same) except for the difference in CUI vs. CUBISTA. The two limiters are more or less the same, with CUBISTA being a bit more accurate. However, it is possible that these two limiters perform differently for this large density case. 

The fix that we are working on is "supposed" to make all 6 cases work. 

In our experience if a particular combination of HO works for the dense bubble case, it would keep working on all problems of interest. For example, you can stick with SSPRK3+CUBISTA for both density and momentum. I'm not sure if WENO will satisfy the CBC criterion, which you need to maintain the bounds of density (and viscosity)?

**Comment by amneetb** [comment]
There could also be some bug in your CUI implementation. Notice that in both cases it is CUI that fails! So the mismatch of the integrators may not be the issue at all.

**Comment by SIMAO75** [comment]
The cubista works for the case of dense sphere, for drop impact, dam break and other problems. I will see if there is a bug in the CUI implementation because I agree cubista and cui are almost the same. 
as far as WENO is concerned, it does not respect the CBC criterion, but it works for this case, and not for others

**Comment by amneetb** [comment]
Here is our implementation of CUI, CUBISTA, MGAMMA, and FBICS
https://github.com/IBAMR/IBAMR/blob/master/src/navier_stokes/fortran/navier_stokes_staggered_helpers2d.f.m4#L866-L1041

You might want to compare with these ones. Of them CUI is the one that is tested the most and is the workhorse. 

The classical gas dynamics limiters WENO, PPM, etc., do not work for high density ratio flows. The maximum density ratio in gas dynamics applications is ~ 7 (with normal shocks at mach number --> \infty)

**Comment by SIMAO75** [comment]
ok I'll inform you, thanks for your help

**Comment by SIMAO75** [comment]
The cubista works for the case of dense sphere, for drop impact, dam break and other problems. I will see if there is a bug in the CUI implementation because I agree cubista and cui are almost the same. 
as far as WENO is concerned, it does not respect the CBC criterion, but it works for this case, and not for others

**Comment by boyceg** [comment]
It sounds like this is not an IBAMR issue and might be better discussed somewhere other than this Issue tracker.

**Comment by amneetb** [comment]
Moved this to the discussion section.

@SIMAO75 I would be interested to know the outcomes of your experiments with the limiters and integrators.
